velox/exec/tests/PartitionedOutputBufferManagerTest.cpp

```diff
@@ -226,21 +266,270 @@ class PartitionedOutputBufferManagerTest : public testing::Test {
+    destinationBuffer.loadData(&buffer, 100);
+    std::atomic<bool> notified{false};
+    destinationBuffer.getData(
+        1'000'000,
+        0,
+        [&](std::vector<std::unique_ptr<folly::IOBuf>> buffers,
+            int64_t sequence) {
+          ASSERT_EQ(buffers.size(), 1);
+          ASSERT_TRUE(buffers[0].get() == nullptr);
+          notified = true;
+        });
+    ASSERT_TRUE(buffer.empty());
+    ASSERT_FALSE(buffer.hasNoMoreData());
+    ASSERT_FALSE(notified);
+    VELOX_ASSERT_THROW(destinationBuffer.maybeLoadData(&buffer), "");
+    buffer.enqueue(nullptr);
+    destinationBuffer.maybeLoadData(&buffer);
+    destinationBuffer.getAndClearNotify().notify();
+    ASSERT_TRUE(notified);
+  }
+
+  {
+    ArbitraryBuffer buffer;
+    int64_t expectedNumBytes{0};
+    for (int i = 0; i < 10; ++i) {
+      auto page = makeSerializedPage(rowType_, 100);
+      expectedNumBytes += page->size();
+      buffer.enqueue(std::move(page));
+    }
+    DestinationBuffer destinationBuffer;
+    destinationBuffer.loadData(&buffer, 1);
+    ASSERT_EQ(
+        buffer.toString(), "[ARBITRARY_BUFFER PAGES[9] NO MORE DATA[false]]");
+    std::atomic<bool> notified{false};
+    int64_t numBytes{0};
+    auto buffers = destinationBuffer.getData(
+        1'000'000'000,
+        0,
+        [&](std::vector<std::unique_ptr<folly::IOBuf>> /*unused*/,
+            int64_t /*unused*/) { notified = true; });
+    for (const auto& buffer : buffers) {
+      numBytes += buffer->length();
+    }
+    ASSERT_GT(numBytes, 0);
+    ASSERT_FALSE(notified);
+    {
+      auto result = destinationBuffer.getAndClearNotify();
+      ASSERT_EQ(result.callback, nullptr);
+      ASSERT_TRUE(result.data.empty());
+      ASSERT_EQ(result.sequence, 0);
+    }
+    auto pages = destinationBuffer.acknowledge(1, false);
+    ASSERT_EQ(pages.size(), 1);
+
+    buffers = destinationBuffer.getData(
+        1'000'000,
+        1,
+        [&](std::vector<std::unique_ptr<folly::IOBuf>> buffers,
+            int64_t sequence) {
+          ASSERT_EQ(sequence, 1);
+          ASSERT_EQ(buffers.size(), 9);
+          for (const auto& buffer : buffers) {
+            numBytes += buffer->length();
+          }
+          notified = true;
+        });
+    ASSERT_TRUE(buffers.empty());
+    ASSERT_FALSE(notified);
+
+    destinationBuffer.maybeLoadData(&buffer);
+    destinationBuffer.getAndClearNotify().notify();
+    ASSERT_TRUE(notified);
+    ASSERT_TRUE(buffer.empty());
+    ASSERT_FALSE(buffer.hasNoMoreData());
+    ASSERT_EQ(numBytes, expectedNumBytes);
+  }
+}
+
+TEST_F(PartitionedOutputBufferManagerTest, basicPartitioned) {
   vector_size_t size = 100;
 
   std::string taskId = "t0";
-  auto task = initializeTask(taskId, rowType, 5, 1);
+  auto task = initializeTask(
+      taskId, rowType_, PartitionedOutputBuffer::Kind::kPartitioned, 5, 1);
+
+  // Partitioned output buffer doesn't allow to update output buffers once
+  // created.
+  VELOX_ASSERT_THROW(
```

Comment:
Can you add error message to make sure it is nice and user-friendly?

Reply:
Updated.